File: taskmanager/src/test/java/com/example/alexeyglushkov/taskmanager/task/SimpleTaskManagerTest.java
package com.example.alexeyglushkov.taskmanager.task;

import android.os.Handler;
import android.os.Looper;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;

/**
 * Created by alexeyglushkov on 23.08.15.
 */

public class SimpleTaskManagerTest {

    private TaskPoolTestSet poolTestSet;
    private TaskManagerTestSet taskManagerTestSet;
    private TaskManager taskManager;

    @Before
    public void setUp() throws Exception {
        taskManager = new SimpleTaskManager(10, new Handler(Looper.myLooper()));
        poolTestSet = new TaskPoolTestSet();
        taskManagerTestSet = new TaskManagerTestSet();

        poolTestSet.before(taskManager);
        taskManagerTestSet.before(taskManager);
    }

    @Test
    public void testSetMaxLoadingTasks() {
        taskManagerTestSet.setMaxLoadingTasks();
    }

    @Test
    public void testGetLoadingTaskCount() {
        taskManagerTestSet.getLoadingTaskCount();
    }

    @Test
    public void testAddTask() {
        taskManagerTestSet.addTask();
    }

    @Test
    public void testAddStartedTask() {
        taskManagerTestSet.addStartedTask();
    }

    @Test
    public void testAddTheSameTaskWithSkipPolicy() {
        taskManagerTestSet.addTheSameTaskWithSkipPolicy();
    }

    @Test
    public void testAddTheSameTaskWithCancelPolicy() {
        taskManagerTestSet.addTheSameTaskWithCancelPolicy();
    }

    @Test
    public void testTaskCallbackCalled() {
        taskManagerTestSet.taskCallbackCalled();
    }

    @Test
    public void testChangedTaskCallbackCalled() {
        taskManagerTestSet.changedTaskCallbackCalled();
    }

    @Test
    public void testTaskWithCancelledImmediatelyCallbackCalledAfterCancel() {
        taskManagerTestSet.taskWithCancelledImmediatelyCallbackCalledAfterCancel();
    }

    @Test
    public void testTaskWithCancelledImmediatelyAndChangedCallbackCalled() {
        taskManagerTestSet.taskWithCancelledImmediatelyAndChangedCallbackCalled();
    }

    @Test
    public void testAddStateListener() {
        taskManagerTestSet.addStateListener();
    }

    @Test
    public void testRemoveStateListener() {
        taskManagerTestSet.removeStateListener();
    }

    @Test
    public void testRemoveTask() {
        taskManagerTestSet.removeTask();
    }

    @Test
    public void testRemoveUnknownTask() {
        taskManagerTestSet.removeUnknownTask();
    }

    @Test
    public void testCheckTaskRemovingAfterFinishing() {
        taskManagerTestSet.checkTaskRemovingAfterFinishing();
    }

    @Test
    public void testGetTaskFromProvider() {
        taskManagerTestSet.getTaskFromProvider();
    }

    @Test
    public void testAddTaskProvider() {
        taskManagerTestSet.addTaskProvider();
    }

    @Test
    public void testAddTaskProvider2() {
        taskManagerTestSet.addTaskProvider2();
    }

    @Test
    public void testAddTaskProviderWithTheSameId() {
        taskManagerTestSet.addTaskProviderWithTheSameId();
    }

    @Test
    public void testRemoveTaskProvider() {
        taskManagerTestSet.removeTaskProvider();
    }

    @Test
    public void testSetTaskExecutor() {
        taskManagerTestSet.setTaskExecutor();
    }

    @Test
    public void testStartImmediately() {
        taskManagerTestSet.startImmediately();
    }

    @Test
    public void testStartImmediatelySkipPolicy() {
        taskManagerTestSet.startImmediatelySkipPolicy();
    }

    @Test
    public void testStartImmediatelySkipPolicyWithFinish() {
        taskManagerTestSet.startImmediatelySkipPolicyWithFinish();
    }

    @Test
    public void testStartImmediatelyFinish() {
        taskManagerTestSet.startImmediatelyFinish();
    }

    @Test
    public void testStartImmediatelyFinishWithChangedCallback() {
        taskManagerTestSet.startImmediatelyFinishWithChangedCallback();
    }

    @Test
    public void testStartImmediatelyCancelWithChangedCallback() {
        taskManagerTestSet.startImmediatelyCancelWithChangedCallback();
    }

    @Test
    public void testSetTaskProviderPriority() {
        taskManagerTestSet.setTaskProviderPriority();
    }

    @Test
    public void testSetGetHandler() {
        taskManagerTestSet.setGetHandler();
    }

    @Test
    public void testGetTaskCount() {
        taskManagerTestSet.getTaskCount();
    }

    @Test
    public void testGetTasks() {
        taskManagerTestSet.getTasks();
    }

    @Test
    public void testSetLimit() {
        taskManagerTestSet.setLimit();
    }

    @Test
    public void testSetLimitRemove() {
        taskManagerTestSet.setLimitRemove();
    }

    // PoolTests

    @Test
    public void testGetTask() {
        poolTestSet.getTask();
    }

    @Test
    public void testGetUnknownTask() {
        poolTestSet.getUnknownTask();
    }

    @Test
    public void testSetGetUserData() {
        poolTestSet.setGetUserData();
    }
}

File: taskmanager/src/test/java/com/example/alexeyglushkov/taskmanager/task/SimpleTaskManagerTest.kt
package com.example.alexeyglushkov.taskmanager.task

import com.example.alexeyglushkov.taskmanager.SimpleTaskManager
import com.example.alexeyglushkov.taskmanager.TaskManager
import com.example.alexeyglushkov.taskmanager.coordinators.TaskManagerCoordinator
import kotlinx.coroutines.test.TestCoroutineDispatcher
import kotlinx.coroutines.test.TestCoroutineExceptionHandler
import kotlinx.coroutines.test.TestCoroutineScope

import org.junit.Before
import org.junit.Test

/**
 * Created by alexeyglushkov on 23.08.15.
 */

class SimpleTaskManagerTest {

    private lateinit var poolTestSet: TaskPoolTestSet
    private lateinit var taskManagerTestSet: TaskManagerTestSet
    private lateinit var taskManager: TaskManager

    @Before
    @Throws(Exception::class)
    fun before() {
        val coordinator: TaskManagerCoordinator = TestTaskManagerCoordinator()
        val exceptionHandler = TestCoroutineExceptionHandler()
        val scope = TestCoroutineScope(exceptionHandler)

        val taskScopeDispatcher = TestCoroutineDispatcher()
        val taskScope = TestCoroutineScope(taskScopeDispatcher)
        taskManager = SimpleTaskManager(coordinator, scope, taskScope)
        poolTestSet = TaskPoolTestSet()
        taskManagerTestSet = TaskManagerTestSet()
        taskManagerTestSet.controller = object : TaskManagerTestSet.TaskManagerController {
            override fun pauseTaskRunning() {
                taskScopeDispatcher.pauseDispatcher()
            }

            override fun resumeTaskRunning() {
                taskScopeDispatcher.resumeDispatcher()
            }
        }

        poolTestSet.before(taskManager)
        taskManagerTestSet.before(taskManager)
    }

    @Test
    fun testGetLoadingTaskCount() {
        taskManagerTestSet.getLoadingTaskCount()
    }

    @Test
    fun testAddTask() {
        taskManagerTestSet.addTask()
    }

    @Test
    fun testAddStartedTask() {
        taskManagerTestSet.addStartedTask()
    }

    @Test
    fun addTaskTwiceWithSameIdWithSkipPolicyAtStart() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithSkipPolicyAtStart()
    }

    @Test
    fun addTaskTwiceWithSameIdWithSkipPolicyAtEnd() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithSkipPolicyAtEnd()
    }

    @Test
    fun addTaskTwiceWithSameIdWithCancelPolicyAtStart() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithCancelPolicyAtStart()
    }

    @Test
    fun addTaskTwiceWithSameIdWithCancelPolicyAtEnd() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithCancelPolicyAtEnd()
    }

    @Test
    fun addTaskTwiceWithSameIdWithAddDependencyPolicy() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithAddDependencyPolicyAtStart()
    }

    @Test
    fun addTaskTwiceWithSameIdWithAddDependencyPolicyAtEnd() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithAddDependencyPolicyAtEnd()
    }

    @Test
    fun addTaskTwiceWithSameIdWithCompletePolicyAtStart() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithCompletePolicyAtStart()
    }

    @Test
    fun addTaskTwiceWithSameIdWithCompletePolicyAtEnd() {
        taskManagerTestSet.addTaskTwiceWithSameIdWithCompletePolicyAtEnd()
    }

    @Test
    fun testTaskCallbackCalled() {
        taskManagerTestSet.taskCallbackCalled()
    }

    @Test
    fun testTaskWithCancelledImmediatelyCallbackCalledAfterCancel() {
        taskManagerTestSet.taskWithCancelledImmediatelyCallbackCalledAfterCancel()
    }

    @Test
    fun testAddStateListener() {
        taskManagerTestSet.addStateListener()
    }

    @Test
    fun testRemoveStateListener() {
        taskManagerTestSet.removeStateListener()
    }

    @Test
    fun testRemoveTask() {
        taskManagerTestSet.removeTask()
    }

    @Test
    fun testRemoveUnknownTask() {
        taskManagerTestSet.removeUnknownTask()
    }

    @Test
    fun testCheckTaskRemovingAfterFinishing() {
        taskManagerTestSet.checkTaskRemovingAfterFinishing()
    }

    @Test
    fun testGetTaskFromProvider() {
        taskManagerTestSet.getTaskFromProvider()
    }

    @Test
    fun testAddTaskProvider() {
        taskManagerTestSet.addTaskProvider()
    }

    @Test
    fun testAddTaskProvider2() {
        taskManagerTestSet.addTaskProvider2()
    }

    @Test
    fun testAddTaskProviderWithTheSameId() {
        taskManagerTestSet.addTaskProviderWithTheSameId()
    }

    @Test
    fun checkDefaultWaitingTaskProvider() {
        taskManagerTestSet.checkDefaultWaitingTaskProvider()
    }

    @Test
    fun testRemoveTaskProvider() {
        taskManagerTestSet.removeTaskProvider()
    }

    @Test
    fun testRemoveTaskProviderWithWaitingTasks() {
        taskManagerTestSet.removeTaskProviderWithWaitingTasks()
    }

    @Test
    fun testStartImmediately() {
        taskManagerTestSet.startImmediately()
    }

    @Test
    fun testStartImmediatelySkipPolicy() {
        taskManagerTestSet.startImmediatelySkipPolicy()
    }

    @Test
    fun testStartImmediatelySkipPolicyWithFinish() {
        taskManagerTestSet.startImmediatelySkipPolicyWithFinish()
    }

    @Test
    fun testStartImmediatelyFinish() {
        taskManagerTestSet.startImmediatelyFinish()
    }

    @Test
    fun testSetTaskProviderPriority() {
        taskManagerTestSet.setTaskProviderPriority()
    }

    @Test
    fun testSetGetScope() {
        taskManagerTestSet.setGetScope()
    }

    @Test
    fun testGetTaskCount() {
        taskManagerTestSet.getTaskCount()
    }

    @Test
    fun testGetTasks() {
        taskManagerTestSet.getTasks()
    }

    @Test
    fun testCancelWaitingTaskFromPool() {
        taskManagerTestSet.cancelWaitingTaskFromPool()
    }

    @Test
    fun testAddTaskFromPoolWhenMaxLoadingTasksIsEnough() {
        taskManagerTestSet.addTaskFromPoolWhenCanLoad()
    }

    @Test
    fun testAddTaskFromPoolWhenMaxLoadingTasksIsNotEnough() {
        taskManagerTestSet.addTaskFromPoolWhenMaxLoadingTasksIsNotEnough()
    }

    @Test
    fun testCancelStartedTaskFromPool() {
        taskManagerTestSet.cancelStartedTaskFromPool()
    }

    // PoolTests

    @Test
    fun testGetTask() {
        poolTestSet.getTask()
    }

    @Test
    fun testGetUnknownTask() {
        poolTestSet.getUnknownTask()
    }

    @Test
    fun testSetGetUserData() {
        poolTestSet.setGetUserData()
    }
}
